### PR TITLE
Hashing affine group elements

### LINKED ISSUE #38381: AffineGroup elements should be hashable
### Problem Description

Currently `AffineGroup` elements do not implement hash and so can not be used as dictionary keys.

Currently the internal variables of `AffineGroupElement` are mutable, but the `matrix` method is cached suggesting that the values are expected to be immutable. This makes it possible to produce (far fetched) bugs where `matrix` returns something incompatible with the current state of the transformation:
```sage
sage: from sage.groups.affine_gps.affine_group import AffineGroup
sage: G = AffineGroup(2, QQ)
sage: G
Affine Group of degree 2 over Rational Field
sage: g = G([[1, 2], [3, 4]], [5,6])
sage: g.matrix()
[1 2|5]
[3 4|6]
[---+-]
[0 0|1]
sage: g.A()[0,0] = 9
sage: g.A()[0,0] = 9
sage: g
      [9 2]     [5]
x |-> [3 4] x + [6]
sage: g.matrix()
[1 2|5]
[3 4|6]
[---+-]
[0 0|1]
sage: hash(g)
TypeError
```

### Proposed Solution

1. Make affine group elements immutable by calling `set_immutable()` on the parameters `_A` and `_b` in the constructor.
2. Implement a hash.

### Alternatives Considered

If mutability is desired (for some reason I'm not aware of), then the `matrix` method should not be cached to avoid the bug above. In this case, I would imagine we'd also want to implement something explaining how to change the values.

### Additional Information

_No response_

### Is there an existing issue for this?

- [X] I have searched the existing issues for a bug report that matches the one I want to file, without success.

### PR DESCRIPTION
**Pull request by tscrim**
Fixes #38381 and makes sure the defining parts of the group element are immutable.

### :memo: Checklist

- [x] The title is concise and informative.
- [x] The description explains in detail what this PR is about.
- [x] I have linked a relevant issue or discussion.
- [x] I have created tests covering the changes.
- [x] I have updated the documentation and checked the documentation preview.

### CODE REVIEW

File: src/sage/groups/affine_gps/group_element.py [hunk 1]
```diff
@@ -138,6 +140,14 @@ def __init__(self, parent, A, b=0, convert=True, check=True):
                 raise TypeError('b must be an element of ' + str(parent.vector_space()))
             parent._element_constructor_check(A, b)
         super().__init__(parent)
+        if not A.is_immutable():
+            from copy import copy
```
Comment (fchapoton) [inline src/sage/groups/affine_gps/group_element.py]: maybe rather import copy globally in the file ?

Comment (tscrim) [reply src/sage/groups/affine_gps/group_element.py]: Indeed, that is better. Done.

File: src/sage/groups/affine_gps/group_element.py [hunk 2]
```diff
@@ -45,6 +45,7 @@
 from sage.structure.element import MultiplicativeGroupElement
 from sage.structure.richcmp import richcmp, richcmp_not_equal
 
+from copy import copy
```
Comment (mkoeppe) [inline src/sage/groups/affine_gps/group_element.py]: imports from stdlib go to the top of the file, before `from sage...` imports

File: src/sage/groups/affine_gps/group_element.py [hunk 3]
```diff
@@ -497,6 +515,27 @@ def _richcmp_(self, other, op):
...
+
+            sage: F = AffineGroup(3, QQ)
+            sage: g = F([1,2,3,4,5,6,7,8,0], [10,11,12])
+            sage: h = F([1,2,3,4,5,6,7,8,0], [10,11,0])
+            sage: hash(g) == hash(h)
+            False
+            sage: hash(g) == hash(copy(g))
+            True
+            sage: f = g * h
+            sage: hash(f) == hash((f.A(), f.b()))
```
Comment (mkoeppe) [inline src/sage/groups/affine_gps/group_element.py]: I don't think it's a good idea to pin this implementation detail in a doctest

**Comment by tscrim** [comment]
I've addressed the reviewer comments (although I disagree with the detail in the doctest; someone who is changing the hash will clearly see the doctest needs to be updated, doctests are not, and should not be, fixed forever).

@fchapoton Could you do the final review here? Should be quick.

File: src/sage/groups/affine_gps/group_element.py [hunk 4]
```diff
@@ -40,12 +40,13 @@
...
 #*****************************************************************************
 
+from copy import copy
+
 from sage.structure.element import Matrix
 from sage.misc.cachefunc import cached_method
 from sage.structure.element import MultiplicativeGroupElement
 from sage.structure.richcmp import richcmp, richcmp_not_equal
 
-
```
Comment (fchapoton) [inline src/sage/groups/affine_gps/group_element.py]: two blank lines here please

**Review by fchapoton — APPROVED** [review]
ok, good to go

**Comment by tscrim** [comment]
Thank you.

**Pull request merged**